### ISSUE TITLE
[RFC]: Add Goodput Metric to Benchmark Serving

### ISSUE THREAD

**Issue by Imss27** [vllm-project/vllm#8782]
### Motivation.

Currently, all metrics vLLM has are more from the perspectives of GenAI Service Providers. 

In order to provide a measurement from the perspectives of GenAI Service Users, we, from [Hao AI Lab](https://hao-ai-lab.github.io/home/#:~:text=Welcome%20to%20the%20UCSD%20Hao%20AI%20Lab%20website!), propose a new user-defined metric, **Goodput**😎.

For more context, see the google docs link in the following section.⬇️

### Proposed Change.

Doc: [Goodput Support Proposal with more details.](https://docs.google.com/document/d/1TcZS-igoZxPrD8rj6_NM-WmCiQwAvP_-JJf2mXZEyZY/edit?usp=sharing)

TL; DR - While doing benchmark, users can define service level objectives and get a **Goodput** result. We also propose adding **Goodput** to the CI benchmark dashboard.

### Feedback Period.

3~5 bussiness days

### CC List.

@zhisbug @youkaichao @KuntaiDu @WoosukKwon @ywang96 

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by ywang96** [comment]
Hey @Imss27 ! Thanks for making this RFC!

I've read the doc and overall it makes sense to me, so please feel free to implement the changes and ping me when the PR is ready for review!

One suggestion I'd make is that I see you're using `inter_token_latency:30` in the example, but you probably want to use TPOT instead since that fits the definition better.

**Comment by Imss27** [comment]
Thanks @ywang96😸! I am trying to understand the problem you are talking about.

My intinial idea is to provide all the SLOs options to users, including all the request-level metric vLLM has right now. So users can define their goodput in a more flexible way.

As for the problem of bundled tokens. I think in some cases, it is just an unavoidable issue due to the intrinsic nature of tokenization. See #8904 for one example I provided.

Anyway, this is great suggestion and let me just work on an initial implementation.🤗

**Comment by ywang96** [comment]
> Thanks @ywang96😸! I am trying to understand the problem you are talking about.
> 
> My initial idea is to provide all the SLOs options to users, including all the request-level metric vLLM has right now. So users can define their goodput in a more flexible way.
> 
> As for the problem of bundled tokens. I think at least in some cases, it is just an unavoidable issue due to the intrinsic nature of tokenization. See #8904 for one example I provided.
> 
> Anyway, this is great suggestion and let me just work on an initial implementation.🤗

Sounds good! I guess my point is around the definition of the SLO for token generation rate: Are we defining on the rate itself, or the user-facing lag between two streamed events?

For example, let's say one backend generates 100 tokens within 1s (so `TPOT=10ms`), but only streams 5 tokens a time back to the user (so `ITL/inter-packet-latency=50ms`). How do you go about defining the SLO for this scenario?

**Comment by njhill** [comment]
Some related discussion here: https://github.com/vllm-project/vllm/issues/6531

**Comment by Imss27** [comment]
Ahh I see. Thanks for the explanation and pointers! @ywang96 @njhill 

For the goodput in our definition, I think only the tracing data at the per-request level matters. Which means the SLOs should also be at the per-request level. To be more accurately, for each request, it considers TTFT, mean_TPOT... and any request level values. For the case you talked about, it seems to be a good idea not to consider ITL, unless the number of bundled tokens is a consistent number and then we can use mean_ITL.
